src/ansi-c/c_typecheck_base.cpp

```diff
@@ -768,14 +768,37 @@ void c_typecheck_baset::typecheck_declaration(
 
       typecheck_symbol(symbol);
 
-      auto check_history_expr = [&](const exprt expr) {
-        disallow_subexpr_by_id(
-          expr, ID_old, CPROVER_PREFIX "old is not allowed in preconditions.");
-        disallow_subexpr_by_id(
-          expr,
-          ID_loop_entry,
-          CPROVER_PREFIX "loop_entry is not allowed in preconditions.");
-      };
+      auto check_history_expr =
+        [&](const exprt expr, std::string &clause_type) {
+          disallow_subexpr_by_id(
+            expr,
+            ID_old,
+            CPROVER_PREFIX "old is not allowed in " + clause_type + ".");
+          disallow_subexpr_by_id(
+            expr,
+            ID_loop_entry,
+            CPROVER_PREFIX "loop_entry is not allowed in " + clause_type + ".");
+        };
+
+      auto check_return_value =
+        [&](const exprt &expr, std::string &clause_type) {
+          const irep_idt id = CPROVER_PREFIX "return_value";
+
+          auto pred = [&](const exprt &expr) {
+            if(!can_cast_expr<symbol_exprt>(expr))
+              return false;
+
+            return to_symbol_expr(expr).get_identifier() == id;
+          };
+
+          if(!has_subexpr(expr, pred))
+            return;
+
+          error().source_location = expr.source_location();
+          error() << id2string(id) + " is not allowed in " << clause_type << "."
+                  << eom;
+          throw 0;
```

Comment:
With #5837 merged it is now preferable to throw `invalid_source_file_exception` in such a case (see that PR for examples). This is to avoid introducing (and eventually getting rid of) `throw 0`.